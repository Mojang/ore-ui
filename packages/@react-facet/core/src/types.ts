export type Immutable = boolean | number | string | undefined | null

export type ObjectWithImmutables = Record<string, Immutable>

export interface EqualityCheck<T> {
  (): (current: T) => boolean
}

/**
 * Possible values a facet can have.
 * Currently functions are not supported
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Value = string | number | boolean | undefined | null | [] | Record<string, any> | bigint

export type ExtractFacetValues<T extends ReadonlyArray<Facet<unknown>>> = {
  [K in keyof T]: T[K] extends Facet<infer V> ? V : never
}

export const FACET_FACTORY = Symbol('facet-factory')

export interface FacetFactory<T> {
  (): Facet<T>
  factory: typeof FACET_FACTORY
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ExcludeFacetFactory<T> = T extends FacetFactory<any> ? never : T

export interface Listener<T> {
  (value: T): void
}

export type Unsubscribe = () => void

export type Observe<T> = (listener: Listener<T>) => Unsubscribe

export interface Facet<T> {
  get: () => Option<T>
  observe: Observe<T>
}

export interface WritableFacet<T> extends Facet<T> {
  set: (value: T) => void
  setWithCallback: (callback: (previousValue: Option<T>) => Option<T>) => void
}

export interface Update<V> {
  (value: V): void
}

export type Setter<V> = (callbackSetter: V | ((previousValue: Option<V>) => Option<V>)) => void

export type Cleanup = Unsubscribe

export type StartSubscription<V> = (update: Update<V>) => Cleanup

export const isFacet = <T>(value: Value | Facet<T>): value is Facet<T> => {
  return (
    value !== null &&
    value !== undefined &&
    (value as Facet<T>).observe !== undefined &&
    (value as Facet<T>).get !== undefined
  )
}

export type FacetProp<T extends Value> = Facet<T> | ExcludeFacetFactory<T>

export interface Effect<T> {
  (value: T): void | Cleanup
}

export const NO_VALUE = Symbol.for('NoValue')

export type NoValue = typeof NO_VALUE

export type Option<T> = NoValue | T

/**
 * Unit of work that can be scheduled within a batch.
 */
export type Task = {
  (): void
  scheduled?: boolean
}

/**
 * Function that when executed will have its Facet updates scheduled.
 */
export type Batch = () => void

/**
 * CSSStyleDeclaration that has FacetDefinition as values and includes Gameface's extended number based properties
 * Based on Gameface's TypeScript definition: https://github.com/Mojang/Mojang_Cohtml/blob/master/src/Scripting/scripted/typescript/ScriptedCSSStyleDeclaration.d.ts
 */
export type FacetCSSStyleDeclaration = {
  alignContent?: FacetProp<string>
  alignItems?: FacetProp<string>
  alignSelf?: FacetProp<string>
  animation?: FacetProp<string>
  animationDelay?: FacetProp<string>
  animationDirection?: FacetProp<string>
  animationDuration?: FacetProp<string>
  animationFillMode?: FacetProp<string>
  animationIterationCount?: FacetProp<string>
  animationName?: FacetProp<string>
  animationPlayState?: FacetProp<string>
  animationTimingFunction?: FacetProp<string>
  background?: FacetProp<string>
  backgroundColor?: FacetProp<string>
  backgroundImage?: FacetProp<string>
  backgroundPosition?: FacetProp<string>
  backgroundRepeat?: FacetProp<string>
  backgroundSize?: FacetProp<string>
  border?: FacetProp<string>
  borderBottom?: FacetProp<string>
  borderBottomColor?: FacetProp<string>
  borderBottomLeftRadius?: FacetProp<string>
  borderBottomRightRadius?: FacetProp<string>
  borderBottomStyle?: FacetProp<string>
  borderBottomWidth?: FacetProp<string>
  borderColor?: FacetProp<string>
  borderImage?: FacetProp<string>
  borderImageOutset?: FacetProp<string>
  borderImageRepeat?: FacetProp<string>
  borderImageSlice?: FacetProp<string>
  borderImageSource?: FacetProp<string>
  borderImageWidth?: FacetProp<string>
  borderLeft?: FacetProp<string>
  borderLeftColor?: FacetProp<string>
  borderLeftStyle?: FacetProp<string>
  borderLeftWidth?: FacetProp<string>
  borderRadius?: FacetProp<string>
  borderRight?: FacetProp<string>
  borderRightColor?: FacetProp<string>
  borderRightStyle?: FacetProp<string>
  borderRightWidth?: FacetProp<string>
  borderStyle?: FacetProp<string>
  borderTop?: FacetProp<string>
  borderTopColor?: FacetProp<string>
  borderTopLeftRadius?: FacetProp<string>
  borderTopRightRadius?: FacetProp<string>
  borderTopStyle?: FacetProp<string>
  borderTopWidth?: FacetProp<string>
  borderWidth?: FacetProp<string>
  bottom?: FacetProp<string>
  bottomPX?: FacetProp<number>
  bottomVH?: FacetProp<number>
  bottomVW?: FacetProp<number>
  boxShadow?: FacetProp<string>
  clipPath?: FacetProp<string>
  color?: FacetProp<string>
  contain?: FacetProp<string>
  cursor?: FacetProp<string>
  display?: FacetProp<'flex' | 'none'>
  filter?: FacetProp<string>
  flex?: FacetProp<string>
  flexDirection?: FacetProp<string>
  flexWrap?: FacetProp<string>
  font?: FacetProp<string>
  fontFamily?: FacetProp<string>
  fontFitMode?: FacetProp<string>
  fontSize?: FacetProp<string>
  fontStyle?: FacetProp<string>
  fontWeight?: FacetProp<string>
  height?: FacetProp<string>
  heightPX?: FacetProp<number>
  heightVH?: FacetProp<number>
  isolation?: FacetProp<string>
  justifyContent?: FacetProp<string>
  left?: FacetProp<string>
  leftPX?: FacetProp<number>
  leftVH?: FacetProp<number>
  leftVW?: FacetProp<number>
  letterSpacing?: FacetProp<string>
  lineHeight?: FacetProp<string>
  margin?: FacetProp<string>
  marginBottom?: FacetProp<string>
  marginLeft?: FacetProp<string>
  marginRight?: FacetProp<string>
  marginTop?: FacetProp<string>
  mask?: FacetProp<string>
  maskClip?: FacetProp<string>
  maskImage?: FacetProp<string>
  maskMode?: FacetProp<string>
  maskPosition?: FacetProp<string>
  maskRepeat?: FacetProp<string>
  maskSize?: FacetProp<string>
  maxHeight?: FacetProp<string>
  maxHeightPX?: FacetProp<number>
  maxHeightVH?: FacetProp<number>
  maxWidth?: FacetProp<string>
  maxWidthPX?: FacetProp<number>
  maxWidthVW?: FacetProp<number>
  minHeight?: FacetProp<string>
  minHeightPX?: FacetProp<number>
  minHeightVH?: FacetProp<number>
  minWidth?: FacetProp<string>
  minWidthPX?: FacetProp<number>
  minWidthVW?: FacetProp<number>
  mixBlendMode?: FacetProp<string>
  opacity?: FacetProp<number>
  overflow?: FacetProp<string>
  overflowWrap?: FacetProp<string>
  overflowX?: FacetProp<string>
  overflowY?: FacetProp<string>
  padding?: FacetProp<string>
  paddingBottom?: FacetProp<string>
  paddingLeft?: FacetProp<string>
  paddingRight?: FacetProp<string>
  paddingTop?: FacetProp<string>
  perspective?: FacetProp<string>
  perspectiveOrigin?: FacetProp<string>
  pointerEvents?: FacetProp<string>
  position?: FacetProp<string>
  right?: FacetProp<string>
  rightPX?: FacetProp<number>
  rightVH?: FacetProp<number>
  rightVW?: FacetProp<number>
  textAlign?: FacetProp<string>
  textOverflow?: FacetProp<string>
  textShadow?: FacetProp<string>
  textTransform?: FacetProp<string>
  top?: FacetProp<string>
  topPX?: FacetProp<number>
  topVH?: FacetProp<number>
  topVW?: FacetProp<number>
  transform?: FacetProp<string>
  transformOrigin?: FacetProp<string>
  transition?: FacetProp<string>
  transitionDelay?: FacetProp<string>
  transitionDuration?: FacetProp<string>
  transitionProperty?: FacetProp<string>
  transitionTimingFunction?: FacetProp<string>
  userSelect?: FacetProp<string>
  visibility?: FacetProp<string>
  whiteSpace?: FacetProp<string>
  width?: FacetProp<string>
  widthPX?: FacetProp<number>
  widthVW?: FacetProp<number>
  zIndex?: FacetProp<number>
}
