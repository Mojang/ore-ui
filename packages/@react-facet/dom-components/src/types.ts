import React from 'react'

export type FocusCallback<T> = (e: React.FocusEvent<T>) => void
export type TouchCallback<T> = (e: React.TouchEvent<T>) => void
export type MouseCallback<T> = (event: React.MouseEvent<T>) => void

export interface PointerEvents<T> {
  onClick?: MouseCallback<T>
  onMouseDown?: MouseCallback<T>
  onMouseUp?: MouseCallback<T>
  onTouchStart?: TouchCallback<T>
  onTouchMove?: TouchCallback<T>
  onTouchEnd?: TouchCallback<T>
  onMouseEnter?: MouseCallback<T>
  onMouseLeave?: MouseCallback<T>
}

export interface FocusEvents<T> {
  onFocus?: FocusCallback<T>
  onBlur?: FocusCallback<T>
}

export type KeyboardCallback<T> = (event: React.KeyboardEvent<T>) => void

export interface KeyboardEvents<T> {
  onKeyPress?: KeyboardCallback<T>
  onKeyDown?: KeyboardCallback<T>
  onKeyUp?: KeyboardCallback<T>
}
