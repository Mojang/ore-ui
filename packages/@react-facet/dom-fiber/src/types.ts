import { FacetProp, Unsubscribe } from '@react-facet/core'
import { FiberRoot, Reconciler } from 'react-reconciler'
import { Key, MutableRefObject, ReactText } from 'react'

export type FacetFiberRoot = FiberRoot

export type TypeHTML =
  | 'fast-a'
  | 'fast-div'
  | 'fast-span'
  | 'fast-p'
  | 'fast-img'
  | 'fast-textarea'
  | 'fast-input'
  | 'fast-text'
  | 'a'
  | 'div'
  | 'p'
  | 'img'
  | 'textarea'
  | 'input'
  | 'style'

export type TypeSVG =
  | 'fast-circle'
  | 'fast-ellipse'
  | 'fast-line'
  | 'fast-path'
  | 'fast-rect'
  | 'fast-svg'
  | 'fast-use'
  | 'fast-polyline'
  | 'fast-polygon'
  | 'fast-linearGradient'
  | 'fast-radialGradient'
  | 'fast-stop'
  | 'fast-svg-text'
  | 'fast-pattern'
  | 'circle'
  | 'ellipse'
  | 'line'
  | 'path'
  | 'rect'
  | 'svg'
  | 'defs'
  | 'g'
  | 'symbol'
  | 'use'
  | 'polyline'
  | 'polygon'
  | 'linearGradient'
  | 'radialGradient'
  | 'stop'
  | 'text'
  | 'pattern'

export type Type = TypeHTML | TypeSVG

export type InputType = 'text' | 'button' | 'password' | 'checkbox' | 'radio' | 'number'

export interface Style {
  [key: string]: FacetProp<string | number | undefined>
}

export interface Callback {
  (): void
}

export interface NoTimeout {
  (): void
}

export type FocusCallback = (e: FocusEvent) => void
export type TouchCallback = (e: TouchEvent) => void
export type MouseCallback = (event: MouseEvent) => void

export interface PointerEvents {
  onClick?: MouseCallback
  onMouseDown?: MouseCallback
  onMouseMove?: MouseCallback
  onMouseUp?: MouseCallback
  onTouchStart?: TouchCallback
  onTouchMove?: TouchCallback
  onTouchEnd?: TouchCallback
  onMouseEnter?: MouseCallback
  onMouseLeave?: MouseCallback
}

export interface FocusEvents {
  onFocus?: FocusCallback
  onBlur?: FocusCallback
}

export type KeyboardCallback = (event: KeyboardEvent) => void

export interface KeyboardEvents {
  onKeyPress?: KeyboardCallback
  onKeyDown?: KeyboardCallback
  onKeyUp?: KeyboardCallback
}

interface StrictReactElement<P = unknown, T extends string = string> {
  type: T
  props: P
  key: Key | null
}

/**
 * More strict type than default React.ReactNode
 *
 * It prevents passing a function as a Node. This allow us to catch accidental passing of facets as children.
 */
export type StrictReactNode = StrictReactElement | ReactText | Array<StrictReactNode> | boolean | null | undefined

export type ElementProps<T> = PointerEvents &
  FocusEvents &
  KeyboardEvents & {
    key?: string | number

    /**
     * More strict children that doesn't support passing facets
     */
    children?: StrictReactNode

    style?: Style
    ref?: React.Ref<T>

    dangerouslySetInnerHTML?: { __html: string }

    className?: FacetProp<string | undefined>
    cx?: FacetProp<string | undefined>
    cy?: FacetProp<string | undefined>
    d?: FacetProp<string | undefined>
    ['data-droppable']?: FacetProp<boolean | undefined>
    ['data-testid']?: FacetProp<string | undefined>
    ['data-x-ray']?: FacetProp<boolean | undefined>
    fill?: FacetProp<string | undefined>
    id?: FacetProp<string | undefined>
    src?: FacetProp<string | undefined>
    height?: FacetProp<string | undefined>
    href?: FacetProp<string | undefined>
    target?: FacetProp<string | undefined>
    autoPlay?: FacetProp<boolean | undefined>
    loop?: FacetProp<boolean | undefined>
    disabled?: FacetProp<boolean | undefined>
    maxLength?: FacetProp<number | undefined>
    r?: FacetProp<string | undefined>
    rx?: FacetProp<string | undefined>
    ry?: FacetProp<string | undefined>
    rows?: FacetProp<number | undefined>
    stroke?: FacetProp<string | undefined>
    strokeWidth?: FacetProp<string | undefined>
    type?: FacetProp<InputType | undefined>
    value?: FacetProp<string | undefined>
    x?: FacetProp<string | undefined>
    x1?: FacetProp<string | undefined>
    x2?: FacetProp<string | undefined>
    width?: FacetProp<string | undefined>
    y?: FacetProp<string | undefined>
    y1?: FacetProp<string | undefined>
    y2?: FacetProp<string | undefined>
    viewBox?: FacetProp<string | undefined>
    xLinkHref?: FacetProp<string | undefined>
    fillOpacity?: FacetProp<string | undefined>
    strokeOpacity?: FacetProp<string | undefined>
    strokeLinecap?: FacetProp<string | undefined>
    strokeLinejoin?: FacetProp<string | undefined>
    points?: FacetProp<string | undefined>
    offset?: FacetProp<string | undefined>
    stopColor?: FacetProp<string | undefined>
    stopOpacity?: FacetProp<string | undefined>
    fontFamily?: FacetProp<string | undefined>
    fontSize?: FacetProp<string | undefined>
  }

export type TextProps = {
  text: FacetProp<string | number>
}

export type Props<T> = ElementProps<T> & TextProps

export type ValidPropsNames = keyof Props<unknown>

export type ElementContainer = {
  children: Set<ElementContainer>

  element: HTMLElement | SVGElement | Text

  style?: CSSStyleDeclaration
  styleUnsubscribers?: Map<string | number, Unsubscribe>

  className?: Unsubscribe
  cx?: Unsubscribe
  cy?: Unsubscribe
  d?: Unsubscribe
  ['data-droppable']?: Unsubscribe
  ['data-testid']?: Unsubscribe
  ['data-x-ray']?: Unsubscribe
  fill?: Unsubscribe
  id?: Unsubscribe
  src?: Unsubscribe
  height?: Unsubscribe
  href?: Unsubscribe
  target?: Unsubscribe
  autoPlay?: Unsubscribe
  loop?: Unsubscribe
  disabled?: Unsubscribe
  maxLength?: Unsubscribe
  r?: Unsubscribe
  rx?: Unsubscribe
  ry?: Unsubscribe
  rows?: Unsubscribe
  stroke?: Unsubscribe
  strokeWidth?: Unsubscribe
  type?: Unsubscribe
  text?: Unsubscribe
  value?: Unsubscribe
  x?: Unsubscribe
  x1?: Unsubscribe
  x2?: Unsubscribe
  width?: Unsubscribe
  y?: Unsubscribe
  y1?: Unsubscribe
  y2?: Unsubscribe
  viewBox?: Unsubscribe
  xLinkHref?: Unsubscribe
  fillOpacity?: Unsubscribe
  strokeOpacity?: Unsubscribe
  strokeLinecap?: Unsubscribe
  strokeLinejoin?: Unsubscribe
  points?: Unsubscribe
  offset?: Unsubscribe
  stopColor?: Unsubscribe
  stopOpacity?: Unsubscribe
  fontFamily?: Unsubscribe
  fontSize?: Unsubscribe
}

export const isElementContainer = (value: ElementContainer | TextContainer): value is ElementContainer => {
  return value != null && 'children' in value
}

export type TextContainer = {
  element: Text
}

export type Instance = ElementContainer
export type Container = Instance
export type TextInstance = TextContainer
export type HydratableInstance = Instance
export type PublicInstance = HTMLElement | SVGElement | Text

export type ReactFacetReconciler = Reconciler<Instance, TextInstance, Container, PublicInstance> & {
  flushPassiveEffects: () => boolean
  IsThisRendererActing: MutableRefObject<boolean>
}

export interface HostContext {
  type?: Type
}

export type UpdatePayload = boolean

export type FastAProps = ElementProps<HTMLAnchorElement>
export type FastCircleProps = ElementProps<SVGCircleElement>
export type FastDivProps = ElementProps<HTMLDivElement>
export type FastEllipseProps = ElementProps<SVGEllipseElement>
export type FastLineProps = ElementProps<SVGLineElement>
export type FastImgProps = ElementProps<HTMLImageElement>
export type FastTextareaProps = ElementProps<HTMLTextAreaElement>
export type FastInputProps = ElementProps<HTMLInputElement>
export type FastPProps = ElementProps<HTMLParagraphElement>
export type FastPathProps = ElementProps<SVGPathElement>
export type FastRectProps = ElementProps<SVGRectElement>
export type FastSpanProps = ElementProps<HTMLSpanElement>
export type FastSvgProps = ElementProps<SVGSVGElement>
export type FastTextProps = TextProps
export type FastUseProps = ElementProps<SVGUseElement>
export type FastPolylineProps = ElementProps<SVGPolylineElement>
export type FastPolyGonProps = ElementProps<SVGPolygonElement>
export type FastLinearGradientProps = ElementProps<SVGLinearGradientElement>
export type FastRadialGradientProps = ElementProps<SVGRadialGradientElement>
export type FastStopProps = ElementProps<SVGStopElement>
export type FastSvgTextProps = ElementProps<SVGTextElement>
export type FastPatternProps = ElementProps<SVGPatternElement>

/**
 * Extends React global namespace with the "fast" types
 */
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    interface IntrinsicElements {
      'fast-a': FastAProps
      'fast-circle': FastCircleProps
      'fast-ellipse': FastEllipseProps
      'fast-line': FastLineProps
      'fast-div': FastDivProps
      'fast-img': FastImgProps
      'fast-textarea': FastTextareaProps
      'fast-input': FastInputProps
      'fast-p': FastPProps
      'fast-path': FastPathProps
      'fast-rect': FastRectProps
      'fast-span': FastSpanProps
      'fast-text': FastTextProps
      'fast-svg': FastSvgProps
      'fast-use': FastUseProps
      'fast-polyline': FastPolylineProps
      'fast-polygon': FastPolyGonProps
      'fast-linearGradient': FastLinearGradientProps
      'fast-radialGradient': FastRadialGradientProps
      'fast-stop': FastStopProps
      'fast-svg-text': FastSvgTextProps
      'fast-pattern': FastPatternProps
    }
  }
}
