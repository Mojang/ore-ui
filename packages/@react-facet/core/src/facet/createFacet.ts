import { defaultEqualityCheck } from '../equalityChecks'
import { Cleanup, EqualityCheck, Listener, WritableFacet, StartSubscription, Option, NO_VALUE } from '../types'
import { batch } from '../scheduler'

export interface FacetOptions<V> {
  initialValue: Option<V>
  startSubscription?: StartSubscription<V>
  equalityCheck?: EqualityCheck<V>
}
/**
 * The low level function to create a Facet, not recommended to be used if you can use any of the react facet hooks to create facets instead (Ex: useFacetState, useFacetWrap)
 */
export function createFacet<V>({
  initialValue,
  startSubscription,
  equalityCheck = defaultEqualityCheck,
}: FacetOptions<V>): WritableFacet<V> {
  const listeners: Set<Listener<V>> = new Set()
  let currentValue = initialValue
  let cleanupSubscription: Cleanup | undefined

  const checker = equalityCheck?.()

  const update = (newValue: V) => {
    if (equalityCheck !== undefined) {
      // we optimize for the most common scenario of using the defaultEqualityCheck (by inline its implementation)
      if (equalityCheck === defaultEqualityCheck) {
        const typeofValue = typeof newValue
        if (
          (typeofValue === 'number' ||
            typeofValue === 'string' ||
            typeofValue === 'boolean' ||
            newValue === null ||
            newValue === undefined) &&
          currentValue === newValue
        ) {
          return
        }
      } else {
        if (checker !== undefined && checker(newValue)) {
          return
        }
      }
    }

    batch(() => {
      currentValue = newValue

      for (const listener of listeners) {
        listener(currentValue)
      }
    })
  }

  /**
   * Simpler update implementation that only resets the value and runs all cleanup functions.
   * Done as a separated function to not interfere with the usual "hot-path" of the update function.
   */
  const updateToNoValue = () => {
    currentValue = NO_VALUE
  }

  return {
    set: update,

    setWithCallback: (setter) => {
      const value = setter(currentValue)
      if (value === NO_VALUE) {
        updateToNoValue()
      } else {
        update(value)
      }
    },

    get: () => currentValue,

    observe: (listener) => {
      listeners.add(listener)

      if (currentValue !== NO_VALUE) {
        listener(currentValue)
      }

      // This is the first subscription, so we start subscribing to dependencies
      if (listeners.size === 1 && startSubscription) {
        cleanupSubscription = startSubscription(update)
      }

      return () => {
        listeners.delete(listener)

        // if this was the last to unsubscribe, we unsubscribe from our dependencies
        if (listeners.size === 0 && cleanupSubscription) {
          currentValue = initialValue
          cleanupSubscription()
        }
      }
    },
  }
}
