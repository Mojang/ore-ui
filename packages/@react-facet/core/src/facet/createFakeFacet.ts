import { Facet } from '..'
/**
 * Creates a nonwritable barebones facet to be used when you need an initial facet value that's meant to be replaced later by a real facet. Ex: with `useContext()`
 */
export function createFakeFacet<T>(value: T): Facet<T> {
  const facet: Facet<T> = {
    get: () => value,
    observe: (listener) => {
      if (process.env.NODE_ENV !== 'production') {
        console.log(`Accessing a fake facet, perhaps you're missing a Context Provider?`)
      }
      listener(value)
      return () => {}
    },
  }
  return facet
}
