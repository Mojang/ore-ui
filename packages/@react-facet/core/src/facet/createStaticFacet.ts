import { Facet } from '../types'
/**
 * Creates a nonwritable barebones static facet to be used when you need an initial facet value outside the react context
 * that's meant to be replaced later by a real facet. Ex: with `createContext()`
 */
export function createStaticFacet<T>(value: T): Facet<T> {
  const facet: Facet<T> = {
    get: () => value,
    observe: (listener) => {
      listener(value)
      return () => {}
    },
  }
  return facet
}
