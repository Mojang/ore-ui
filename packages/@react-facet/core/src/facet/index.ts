export * from './createFacet'
export * from './createFacet'
