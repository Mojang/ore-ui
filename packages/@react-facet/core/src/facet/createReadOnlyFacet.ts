import { Facet } from '../types'
import { createFacet, FacetOptions } from './createFacet'

export function createReadOnlyFacet<V>(options: FacetOptions<V>): Facet<V> {
  return createFacet(options)
}
