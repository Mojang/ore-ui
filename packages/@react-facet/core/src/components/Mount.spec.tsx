import React from 'react'
import { render } from '@react-facet/dom-fiber-testing-library'
import { Mount } from '.'
import { createFacet } from '../facet'
import { NO_VALUE } from '../types'

it('renders when true', () => {
  const display = createFacet({ initialValue: true })
  const rendered = jest.fn()

  const Content = () => {
    rendered()
    return <div>Hello there</div>
  }

  const Example = () => {
    return (
      <Mount when={display}>
        <Content />
      </Mount>
    )
  }

  const scenario = <Example />

  const { container } = render(scenario)

  expect(rendered).toHaveBeenCalled()
  expect(container.firstChild).toMatchSnapshot()
})

it('does not render when false', () => {
  const display = createFacet({ initialValue: false })
  const rendered = jest.fn()

  const Content = () => {
    rendered()
    return <div>Hello there</div>
  }

  const Example = () => {
    return (
      <Mount when={display}>
        <Content />
      </Mount>
    )
  }

  const scenario = <Example />

  render(scenario)

  expect(rendered).not.toHaveBeenCalled()
})

it('does not render facet has no value', () => {
  const display = createFacet<boolean>({ initialValue: NO_VALUE })
  const rendered = jest.fn()

  const Content = () => {
    rendered()
    return <div>Hello there</div>
  }

  const Example = () => {
    return (
      <Mount when={display}>
        <Content />
      </Mount>
    )
  }

  const scenario = <Example />

  render(scenario)

  expect(rendered).not.toHaveBeenCalled()
})
