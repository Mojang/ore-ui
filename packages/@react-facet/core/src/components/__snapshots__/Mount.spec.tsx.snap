// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when true 1`] = `
<div>
  Hello there
</div>
`;
